kanidmd/src/lib/credential/mod.rs

```
@@ -706,6 +768,59 @@ impl Credential {
         }
     }
 
+    pub(crate) fn update_backup_code(&self, backup_code: HashSet<String>) -> Self {
+        let type_ = match &self.type_ {
+            // TODO: we should panic/reject CredentialType != PasswordMfa?
+            CredentialType::Password(pw) | CredentialType::GeneratedPassword(pw) => {
+                debug_assert!(false);
+                CredentialType::PasswordMfa(
+                    pw.clone(),
+                    None,
+                    Map::new(),
+                    Some(BackupCode {
+                        code_set: backup_code,
+                    }),
+                )
+            }
+            CredentialType::PasswordMfa(pw, totp, wan, _) => CredentialType::PasswordMfa(
+                pw.clone(),
+                totp.clone(),
+                wan.clone(),
+                Some(BackupCode {
+                    code_set: backup_code,
+                }),
+            ),
+            CredentialType::Webauthn(wan) => {
+                debug_assert!(false);
```

Should return a Err() instead.